### ISSUE TITLE
Method "size" should be removed from Reservoir interface

### ISSUE THREAD

**Issue by vladimir-bukhtoyarov** [dropwizard/metrics#874]
At the moment, I am doing integration of metrics-core and HdrHistogram, and I am unable to understand what I should return as "size". Method looks as unnecessary due to never used outside concrete Reservoir implementations.

**Comment by ryantenney** [comment]
Reservoir Javadocs specify that it is to return the number of values recorded. Absent a compelling reason to remove it I'd like to leave well enough alone.

```java
    /**
     * Returns the number of values recorded.
     *
     * @return the number of values recorded
     */
    int size();
```

**Comment by vladimir-bukhtoyarov** [comment]
What is mean "recorded"? If I call instance of ExponentiallyDecayingReservoir(configured with size 1028) by 1000000 times, what is correctly value of size should be 1028 or 1000000?

**Comment by vladimir-bukhtoyarov** [comment]
In my implementation [HdrReservoir](https://github.com/vladimir-bukhtoyarov/metrics-core-hdr/blob/master/src/main/java/com/github/metricscore/hdrhistogram/HdrReservoir.java),  I decided to disallow user to call this method, because nobody knows what is it and how it should work.

**Comment by marshallpierce** [comment]
I already have an HdrHistogram-backed reservoir implementation, btw. https://bitbucket.org/marshallpierce/hdrhistogram-metrics-reservoir may do what you need.

**Comment by vladimir-bukhtoyarov** [comment]
Thanks Marshall, 
I sow your library, moreover I start my project from it. But your solution does not cover several corner cases which may be not interested for guys which use "push" as way to dilevery metrics to monitoring database. But my company uses Zabbix as monitoring system and "pull" delivery model. When I say "pull" I mean that monitoring system ask applications about their metrics. I understand that "pull" is bad way and modern databases like influxdb and graphite uses "push" model when application writes metrics to monitoring database, but I can not to do nothing with this, so Zabbix Java Proxy to collecting metrics. And your solution absolutely unsuitable for using with the Zabbix Java Proxy because:

1. HdrHistogramResetOnSnapshotReservoir can not be used in case when more than one parameter of histogram should be saved in monitoring database. For example I want to store in Zabbix 95percentile and 95percentile. Zabbix proxy ask JVM via JMX about 96pecentile, snapshot taken in this point and histogram reset, then proxy ask 99percentile via JMX but histogram already cleared in previous step, so 99percentile is wrong.
2. Snapshot produces amount garbage which equals to histogram size, because you copy histogram each time when snapshot taken. But if I already know list of percentiles which I want to persist in monitoring database then I can copy only several bytes from source histogram and avoid full coping.

I do not want to say that your library is bad, it is just not suitable for my cases. Because of this I decide to create own solution. Additionally I covered many another problems which not so critical but nice to have to be fixed.

And current issue #874 is only one problem which I have with my library. I do not understand why fucking "size" method is part of public API, there are no reasons to place it to Reservoir interface. I sow that you had similar problem with understanding of this method when create comment to [![your code](https://bitbucket.org/marshallpierce/hdrhistogram-metrics-reservoir/src/9137f6b45aa98ec1ed2bfddd9a8b48f7ebf88d60/src/main/java/org/mpierce/metrics/reservoir/hdrhistogram/HdrHistogramReservoir.java?at=master&fileviewer=file-view-default):
 ```java
@Override
    public int size() {
        // This appears to be infrequently called, so not keeping a separate counter just for this.
        return getSnapshot().size();
    }
 ```

**Comment by marshallpierce** [comment]
Ah yes, the pull model is quite problematic... You may find it easier to simply build your own Reporter that stores data and then expose data from there to Zabbix. This would make the allocation less painful, and mean you wouldn't have to change Reservoir semantics to support that. You could just use any old reservoir you like and implement your own custom stuff separately from what Zabbix forces on you.

Anyway, one of my goals for v4 is a completely zero-allocation reporting pipeline, so let's hope in the future this sort of thing will be easier. :)

**Comment by vladimir-bukhtoyarov** [comment]
@marshallpierce 

Sorry for trouble. When you say "reporting pipeline" do you mean some new concept which will be introduced to metrics-core v4 at the library level as first class citizen? Currently the lack of mediator between primary metrics and exporters leads to fundamental problem to show each measure to exporter once and only once, and in same time do not miss measures in case of many reporter. Of course this sort of problem is not actual for built-in to metrics-core reservoir implementations because all of its have a sampling nature and has no goals to achieve at most once delivery semantic, but sampling this is a very big problem by itself which leads to reporting obsolete measures at low event rate and potentially loses min and max measures at high event rate.

**Comment by marshallpierce** [comment]
Yeah, I agree those are real problems -- my thinking is currently that all metrics should have their data extracted as deltas according to some schedule which are then made available to all registered reporters. But, in the meantime, could you not hack together a Reporter that does this under the hood and caches data for a short while so that when Zabbix asks for data X and Y separately, those data points could come from the same underlying set of data?
